docs/_posts/2022-08-27-zero-inference.md

```diff
@@ -0,0 +1,100 @@
+
+### Optimizations
+To further improve system efficiency, ZeRO-Inference leverages two additional optimizations to reduce the latency of fetching layer weights from CPU or NVMe memory into GPU memory. The first optimization involves overlapping the fetch of a particular layer with the computation of an earlier layer, a.k.a., layer prefetching. With layer prefetching, ZeRO-Inference can hide portions of the transfer latency of prefetched layers. Layer prefetching is especially effective for computations on long sequence lengths and/or large batch sizes. The second optimization, which is applicable for inference on multiple GPUs, involves parallelizing the fetch of each layer across multiple GPUs by using each GPU to fetch only a portion of the layer. Employing the aggregate PCIe links of the GPUs in this manner essentially increases the transfer bandwidth linearly, thus reducing the latency. With this approach, fetching layers into GPU memory occurs in two phases. First, each GPU independently fetches a partition of the layer over PCIe into its memory. At this point, only a partition of the layer will be resident on each GPU.  Next, each GPU assembles the full layer for computation by fetching the missing layer pieces from other GPUs over the high-bandwidth GPU-GPU interconnect (e.g., NVLink, xGMI, etc.). Since GPU-GPU interconnect bandwidth is typically over an order of magnitude higher than PCIe bandwidth, efficient multi-GPU or multi-node communication primitives, such as NCCL or RCCL all-gather, can be used to efficiently assemble the full layer on all GPUs with negligible latency compared to the PCIe latency.
+
+### Alternative approach: Host some model weights in GPU memory
+An alternative approach to ZeRO-Inference is to pin as much of the model weights as possible into GPU memory and fetch the remainder (from CPU or NVMe) when needed for computation. A benefit of this approach is avoidance of the latency of fetching weights that are already pinned in GPU memory. However, this approach has two downsides: (i) the latency savings for hundred-billion parameter models are negligible since only a small fraction of the weights can fit in GPU memory, and (ii) even when a decent portion of the model weights (e.g., > 50%) can fit (e.g., smaller models with tens-of-billion parameters), the remaining GPU memory only supports small batch sizes which hurts inference throughput. We later show evaluation results to demonstrate that this approach is sub-optimal.
```

Comment:
```suggestion
An alternative approach to ZeRO-Inference is to pin as many of the model weights as possible into GPU memory and fetch the remainder (from CPU or NVMe) when needed for computation. A benefit of this approach is avoidance of the latency of fetching weights that are already pinned in GPU memory. However, this approach has two downsides: (i) the latency savings for hundred-billion parameter models are negligible since only a small fraction of the weights can fit in GPU memory, and (ii) even when a decent portion of the model weights (e.g., > 50%) can fit (e.g., smaller models with tens-of-billion parameters), the remaining GPU memory can only fit small batch sizes which hurts inference throughput. We later show evaluation results to demonstrate that this approach is sub-optimal.
```

Reply:
I'd also massage those double "e.g." in a row - doesn't read easy - I'd combine into a single e.g.:

> even when a decent portion of the model weights can fit (e.g., >50% for ~10B models), the remaining GPU memory ...

Reply:
Done